rfcs/ion_templates.md

```diff
@@ -0,0 +1,472 @@
+      occupation: SDE
+    }, {
+      name: "David",
+      employeeId: 73328,
+      occupation: SDE
+    }, {
+      name: "Wes",
+      employeeId: 19384,
+      occupation: SDE
+    }, 
+  ]
+}
+{
+  name: "Zack",
+  employeeId: 12345,
+  occupation: "SDE",
+  peers: [{
+      name: "Zack",
+      employeeId: 12345,
+      occupation: SDE
+    }, {
+      name: "Fernando",
+      employeeId: 45673,
+      occupation: SDE
+    }, {
+      name: "Peter",
+      employeeId: 23678,
+      occupation: SDE
+    }, {
+      name: "Tyler",
+      employeeId: 11133,
+      occupation: SDE
+    }, {
+      name: "David",
+      employeeId: 73328,
+      occupation: SDE
+    }, {
+      name: "Wes",
+      employeeId: 19384,
+      occupation: SDE
+    }, 
+  ]
+}
+{
+  name: "Tyler",
+  employeeId: 11133,
+  occupation: "SDE",
+  peers: [{
+      name: "Zack",
+      employeeId: 12345,
+      occupation: SDE
+    }, {
+      name: "Fernando",
+      employeeId: 45673,
+      occupation: SDE
+    }, {
+      name: "Peter",
+      employeeId: 23678,
+      occupation: SDE
+    }, {
+      name: "Tyler",
+      employeeId: 11133,
+      occupation: SDE
+    }, {
+      name: "David",
+      employeeId: 73328,
+      occupation: SDE
+    }, {
+      name: "Wes",
+      employeeId: 19384,
+      occupation: SDE
+    }, 
+  ]
+}
+```
+
+## ‘System’ Templates
+
+Ion 1.0 provides [a ‘system’ symbol table](http://amzn.github.io/ion-docs/docs/symbols.html#system-symbols) with 9 pre-defined symbols. Ion 1.1 can introduce a system template table. This table could provide template definitions for common but surprisingly expensive expressions. Examples follow.
+
+### Local Symbol/Template table appends
+
+In its least expensive form, an LST append definition for a new symbol `new_symbol` looks like this:
+
+```js
+// Annotation type descriptor: 1 byte
+// Annotations length: 1 byte
+// Annotation #1 length: 1 byte
+// Annotation #1 symbol ID: 1 byte
+$ion_symbol_table::
+// Struct type descriptor: 1 byte
+// Struct length: 1 byte
+{
+  // Field name: 1 byte
+  // Symbol type descriptor: 1 byte
+  // Symbol: 1 byte
+  imports: $ion_symbol_table,
+  // Field name: 1 byte
+  // List type descriptor: 1 byte
+  // Symbol string: 11 bytes
+  symbols: ["new_symbol"]
+}
+
+// Total bytes: 22 bytes
+// Overhead: **10 bytes**
+```
+
+
+While the list of strings to intern (`["new symbol"]`) is itself only 12 bytes, we have to pay an additional 10 bytes of overhead to add it to the symbol table.
+
+Using a macro invocation like this:
+
+```js
+// Macro type descriptor: 1 byte
+// Macro ID: 1 byte
+// List type descriptor: 1 byte
+// Symbol string: 11 bytes
+@0(["new_symbol"])
+
+// Total bytes: 14 bytes
+// Overhead: **2 bytes**
+```
+
+we can cut the overhead by 80%.
+
+The same technique could be used for local template table appends.
+
+## Shared Templates
+
+Shared template table imports are handled in the same way as shared symbol table imports.
+
```

Comment:
I think this is a bit light on details on how template tables exist relative to their symbol table counterparts.  I really think this specification needs to spell this out better.  For example, do template tables live in a distinct namespace from symbol tables (i.e. the template table named `foo` is distinct from and has *no* relation to the symbol table named `foo`)?

This is one of the reasons why I prefer adding templates to symbol tables, it inherits the logic, including resolution under the same specification, which streamlines this specification and possibly makes implementing this in existing implementations easier (no need to create parallel APIs for template tables, for example).